README
Ogg/Vorbis/Header version 0.04
============================

This module presents an object-oriented interface to Ogg Vorbis files
which allows user to view Vorbis info and comments and to modify or
add comments.  

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

Inline::C, libogg, libvorbis, libogg-dev, libvorbis-dev.

Inline::C is available from CPAN.  The libogg and libvorbis libraries
are available from xiph.org.  They are also available through Debian's
apt-get and Mandrake's rpmdrake (and perhaps Redhat's up2date).

COPYRIGHT AND LICENCE

Copyright (c) 2002, Dan Pemstein.  All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at
your option) any later version.  A copy of this license is included
with this module (LICENSE.GPL).

A library for editing Ogg Vorbis comments is distributed with this
library as unmodified source code (inc/vcedit.h, inc/vcedit.c,
inc/i18n.h).  This library is Copyright (c) Michael Smith
<msmith@labyrinth.net.au>.  It is licensed under the GNU Library
General Public License (LGPL).  A copy of this license is included
with this module (inc/LICENSE.LGPL).


Changes
Revision history for Perl extension Ogg::Vorbis::Header.

0.01  Tue Dec 31 05:50:10 2002
	- original version; created by h2xs 1.21 with options
		-PAXn Ogg::Vorbis::Header
0.02  Sun Jan  5 04:49:33 CST 2003
  - write_vorbis() no longer croaks on files with spaces in their
    names.
  - Replaced system kludge in write_vorbis() with explicit copy.
0.03  Wed May 21 05:13:25 CDT 2003
  - Added fix for non-standard comments which don't contain '='
		symbols.  Library issues a warning and skips these comments when
		loading from file.

0.04  Fri Jul 01 19:07:00 BST 2016
  - Fixed syntax errors in tests.

0.05  Sun Jul 03 06:25:00 BST 2016
  - Packaging fixes

0.06  Sun Mar 04 10:45:00 GMT 2018
  - Added repo details to metadata

0.10  Mon Mar 19 15:30:00 GMT 2018
  - Move to a modern distribution layout


test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

#########################

# change 'tests => 1' to 'tests => last_test_to_print';

use strict;
use warnings;

use Test;
BEGIN { plan tests => 13 }
use Ogg::Vorbis::Header;
ok(1); # If we made it this far, we're ok.

#########################

# Insert your test code below, the Test module is use()ed here so read
# its man page ( perldoc Test ) for help writing this test script.


# See if partial load works
ok(my $ogg = Ogg::Vorbis::Header->new("test.ogg"));

# See if load_after works
ok($ogg->load);

# Try all the routines
ok($ogg->info->{"rate"} == 44_100);
ok($ogg->comment_tags);
ok(($ogg->comment("artist"))[0] eq "Dan");
ok($ogg->add_comments("moog", "bog"));
ok($ogg->edit_comment("moog", "bug"));
ok($ogg->delete_comment("artist"));
ok($ogg->write_vorbis);
system("cp test.ogg test.ogg.2");
system("cp test.ogg.bak test.ogg");
ok($ogg->clear_comments);

# See if full load works
ok(my $ogg = Ogg::Vorbis::Header->load("test.ogg.2"));
ok(($ogg->comment("moog"))[0] eq "bug");

unlink("test.ogg.2");


Makefile.PL
use Inline::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    'NAME'         => 'Ogg::Vorbis::Header',
    'VERSION_FROM' => 'lib/Ogg/Vorbis/Header.pm', # finds $VERSION
    'LICENSE'      => 'gpl_2',
    'PREREQ_PM'    => {
       Inline::C   => 0,
    }, # e.g., Module::Name => 1.1
    'CONFIGURE_REQUIRES' => {
       Inline::MakeMaker => 0,
    },
    META_MERGE => {
       'meta-spec' => { version => 2 },
       dynamic_config => 0,
      resources => {
        repository => {
          type => 'git',
          url  => 'https://github.com/davorg/Ogg-Vorbis-Header.git',
          web  => 'https://github.com/davorg/Ogg-Vorbis-Header',
        },
      },
    },
    ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM => 'lib/Ogg/Vorbis/Header.pm', # retrieve abstract from module
       AUTHOR     => 'Dan Pemstein <dan@lcws.org>') : ()),
);


inc/vcedit.c
/* This program is licensed under the GNU Library General Public License, version 2,
 * a copy of which is included with this program (LICENCE.LGPL).
 *
 * (c) 2000-2001 Michael Smith <msmith@labyrinth.net.au>
 *
 *
 * Comment editing backend, suitable for use by nice frontend interfaces.
 *
 * last modified: $Id: vcedit.c,v 1.1.1.1 2003/01/05 00:16:43 dan Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <ogg/ogg.h>
#include <vorbis/codec.h>

#include "vcedit.h"
#include "i18n.h"


#define CHUNKSIZE 4096

vcedit_state *vcedit_new_state(void)
{
  vcedit_state *state = malloc(sizeof(vcedit_state));
  memset(state, 0, sizeof(vcedit_state));

  return state;
}

char *vcedit_error(vcedit_state *state)
{
  return state->lasterror;
}

vorbis_comment *vcedit_comments(vcedit_state *state)
{
  return state->vc;
}

static void vcedit_clear_internals(vcedit_state *state)
{
  if(state->vc)
  {
    vorbis_comment_clear(state->vc);
    free(state->vc);
  }
  if(state->os)
  {
    ogg_stream_clear(state->os);
    free(state->os);
  }
  if(state->oy)
  {
    ogg_sync_clear(state->oy);
    free(state->oy);
  }
  if(state->vendor)
    free(state->vendor);
    if(state->mainbuf)
        free(state->mainbuf);
    if(state->bookbuf)
        free(state->bookbuf);
    if(state->vi) {
       	vorbis_info_clear(state->vi);
        free(state->vi);
    }

    memset(state, 0, sizeof(*state));
}

void vcedit_clear(vcedit_state *state)
{
  if(state)
  {
    vcedit_clear_internals(state);
    free(state);
  }
}

/* Next two functions pulled straight from libvorbis, apart from one change
 * - we don't want to overwrite the vendor string.
 */
static void _v_writestring(oggpack_buffer *o,char *s, int len)
{
  while(len--)
  {
    oggpack_write(o,*s++,8);
  }
}

static int _commentheader_out(vorbis_comment *vc, char *vendor, ogg_packet *op)
{
  oggpack_buffer opb;

  oggpack_writeinit(&opb);

  /* preamble */
  oggpack_write(&opb,0x03,8);
  _v_writestring(&opb,"vorbis", 6);

  /* vendor */
  oggpack_write(&opb,strlen(vendor),32);
  _v_writestring(&opb,vendor, strlen(vendor));

  /* comments */
  oggpack_write(&opb,vc->comments,32);
  if(vc->comments){
    int i;
    for(i=0;i<vc->comments;i++){
      if(vc->user_comments[i]){
        oggpack_write(&opb,vc->comment_lengths[i],32);
        _v_writestring(&opb,vc->user_comments[i],
                        vc->comment_lengths[i]);
      }else{
        oggpack_write(&opb,0,32);
      }
    }
  }
  oggpack_write(&opb,1,1);

  op->packet = _ogg_malloc(oggpack_bytes(&opb));
  memcpy(op->packet, opb.buffer, oggpack_bytes(&opb));

  op->bytes=oggpack_bytes(&opb);
  op->b_o_s=0;
  op->e_o_s=0;
  op->granulepos=0;

  oggpack_writeclear(&opb);
  return 0;
}

static int _blocksize(vcedit_state *s, ogg_packet *p)
{
  int this = vorbis_packet_blocksize(s->vi, p);
  int ret = (this + s->prevW)/4;

  if(!s->prevW)
  {
    s->prevW = this;
    return 0;
  }

  s->prevW = this;
  return ret;
}

static int _fetch_next_packet(vcedit_state *s, ogg_packet *p, ogg_page *page)
{
  int result;
  char *buffer;
  int bytes;

  result = ogg_stream_packetout(s->os, p);

  if(result > 0)
    return 1;
  else
  {
    if(s->eosin)
      return 0;
    while(ogg_sync_pageout(s->oy, page) <= 0)
    {
      buffer = ogg_sync_buffer(s->oy, CHUNKSIZE);
      bytes = s->read(buffer,1, CHUNKSIZE, s->in);
      ogg_sync_wrote(s->oy, bytes);
      if(bytes == 0)
        return 0;
    }
    if(ogg_page_eos(page))
      s->eosin = 1;
    else if(ogg_page_serialno(page) != s->serial)
    {
      s->eosin = 1;
      s->extrapage = 1;
      return 0;
    }

    ogg_stream_pagein(s->os, page);
    return _fetch_next_packet(s, p, page);
  }
}

int vcedit_open(vcedit_state *state, FILE *in)
{
  return vcedit_open_callbacks(state, (void *)in,
      (vcedit_read_func)fread, (vcedit_write_func)fwrite);
}

int vcedit_open_callbacks(vcedit_state *state, void *in,
    vcedit_read_func read_func, vcedit_write_func write_func)
{

  char *buffer;
  int bytes,i;
  ogg_packet *header;
  ogg_packet	header_main;
  ogg_packet  header_comments;
  ogg_packet	header_codebooks;
  ogg_page    og;

  state->in = in;
  state->read = read_func;
  state->write = write_func;

  state->oy = malloc(sizeof(ogg_sync_state));
  ogg_sync_init(state->oy);

  buffer = ogg_sync_buffer(state->oy, CHUNKSIZE);
  bytes = state->read(buffer, 1, CHUNKSIZE, state->in);

  ogg_sync_wrote(state->oy, bytes);

  if(ogg_sync_pageout(state->oy, &og) != 1)
  {
    if(bytes<CHUNKSIZE)
      state->lasterror = _("Input truncated or empty.");
    else
      state->lasterror = _("Input is not an Ogg bitstream.");
    goto err;
  }

  state->serial = ogg_page_serialno(&og);

  state->os = malloc(sizeof(ogg_stream_state));
  ogg_stream_init(state->os, state->serial);

    state->vi = malloc(sizeof(vorbis_info));
  vorbis_info_init(state->vi);

  state->vc = malloc(sizeof(vorbis_comment));
  vorbis_comment_init(state->vc);

  if(ogg_stream_pagein(state->os, &og) < 0)
  {
    state->lasterror = _("Error reading first page of Ogg bitstream.");
    goto err;
  }

  if(ogg_stream_packetout(state->os, &header_main) != 1)
  {
    state->lasterror = _("Error reading initial header packet.");
    goto err;
  }

  if(vorbis_synthesis_headerin(state->vi, state->vc, &header_main) < 0)
  {
    state->lasterror = _("Ogg bitstream does not contain vorbis data.");
    goto err;
  }

  state->mainlen = header_main.bytes;
  state->mainbuf = malloc(state->mainlen);
  memcpy(state->mainbuf, header_main.packet, header_main.bytes);

  i = 0;
  header = &header_comments;
  while(i<2) {
    while(i<2) {
      int result = ogg_sync_pageout(state->oy, &og);
      if(result == 0) break; /* Too little data so far */
      else if(result == 1)
      {
        ogg_stream_pagein(state->os, &og);
        while(i<2)
        {
          result = ogg_stream_packetout(state->os, header);
          if(result == 0) break;
          if(result == -1)
          {
            state->lasterror = _("Corrupt secondary header.");
            goto err;
          }
          vorbis_synthesis_headerin(state->vi, state->vc, header);
          if(i==1)
          {
            state->booklen = header->bytes;
            state->bookbuf = malloc(state->booklen);
            memcpy(state->bookbuf, header->packet,
                header->bytes);
          }
          i++;
          header = &header_codebooks;
        }
      }
    }

    buffer = ogg_sync_buffer(state->oy, CHUNKSIZE);
    bytes = state->read(buffer, 1, CHUNKSIZE, state->in);
    if(bytes == 0 && i < 2)
    {
      state->lasterror = _("EOF before end of vorbis headers.");
      goto err;
    }
    ogg_sync_wrote(state->oy, bytes);
  }

  /* Copy the vendor tag */
  state->vendor = malloc(strlen(state->vc->vendor) +1);
  strcpy(state->vendor, state->vc->vendor);

  /* Headers are done! */
  return 0;

err:
  vcedit_clear_internals(state);
  return -1;
}

int vcedit_write(vcedit_state *state, void *out)
{
  ogg_stream_state streamout;
  ogg_packet header_main;
  ogg_packet header_comments;
  ogg_packet header_codebooks;

  ogg_page ogout, ogin;
  ogg_packet op;
  ogg_int64_t granpos = 0;
  int result;
  char *buffer;
  int bytes;
  int needflush=0, needout=0;

  state->eosin = 0;
  state->extrapage = 0;

  header_main.bytes = state->mainlen;
  header_main.packet = state->mainbuf;
  header_main.b_o_s = 1;
  header_main.e_o_s = 0;
  header_main.granulepos = 0;

  header_codebooks.bytes = state->booklen;
  header_codebooks.packet = state->bookbuf;
  header_codebooks.b_o_s = 0;
  header_codebooks.e_o_s = 0;
  header_codebooks.granulepos = 0;

  ogg_stream_init(&streamout, state->serial);

  _commentheader_out(state->vc, state->vendor, &header_comments);

  ogg_stream_packetin(&streamout, &header_main);
  ogg_stream_packetin(&streamout, &header_comments);
  ogg_stream_packetin(&streamout, &header_codebooks);

  while((result = ogg_stream_flush(&streamout, &ogout)))
  {
    if(state->write(ogout.header,1,ogout.header_len, out) !=
        (size_t) ogout.header_len)
      goto cleanup;
    if(state->write(ogout.body,1,ogout.body_len, out) !=
        (size_t) ogout.body_len)
      goto cleanup;
  }

  while(_fetch_next_packet(state, &op, &ogin))
  {
    int size;
    size = _blocksize(state, &op);
    granpos += size;

    if(needflush)
    {
      if(ogg_stream_flush(&streamout, &ogout))
      {
        if(state->write(ogout.header,1,ogout.header_len,
              out) != (size_t) ogout.header_len)
          goto cleanup;
        if(state->write(ogout.body,1,ogout.body_len,
              out) != (size_t) ogout.body_len)
          goto cleanup;
      }
    }
    else if(needout)
    {
      if(ogg_stream_pageout(&streamout, &ogout))
      {
        if(state->write(ogout.header,1,ogout.header_len,
              out) != (size_t) ogout.header_len)
          goto cleanup;
        if(state->write(ogout.body,1,ogout.body_len,
              out) != (size_t) ogout.body_len)
          goto cleanup;
      }
    }

    needflush=needout=0;

    if(op.granulepos == -1)
    {
      op.granulepos = granpos;
      ogg_stream_packetin(&streamout, &op);
    }
    else /* granulepos is set, validly. Use it, and force a flush to
        account for shortened blocks (vcut) when appropriate */
    {
      if(granpos > op.granulepos)
      {
        granpos = op.granulepos;
        ogg_stream_packetin(&streamout, &op);
        needflush=1;
      }
      else
      {
        ogg_stream_packetin(&streamout, &op);
        needout=1;
      }
    }
  }

  streamout.e_o_s = 1;
  while(ogg_stream_flush(&streamout, &ogout))
  {
    if(state->write(ogout.header,1,ogout.header_len,
          out) != (size_t) ogout.header_len)
      goto cleanup;
    if(state->write(ogout.body,1,ogout.body_len,
          out) != (size_t) ogout.body_len)
      goto cleanup;
  }

  if (state->extrapage)
  {
    if(state->write(ogin.header,1,ogin.header_len,
                    out) != (size_t) ogin.header_len)
      goto cleanup;
    if (state->write(ogin.body,1,ogin.body_len, out) !=
        (size_t) ogin.body_len)
      goto cleanup;
  }

  state->eosin=0; /* clear it, because not all paths to here do */
  while(!state->eosin) /* We reached eos, not eof */
  {
    /* We copy the rest of the stream (other logical streams)
     * through, a page at a time. */
    while(1)
    {
      result = ogg_sync_pageout(state->oy, &ogout);
      if(result==0)
                break;
      if(result<0)
        state->lasterror = _("Corrupt or missing data, continuing...");
      else
      {
        /* Don't bother going through the rest, we can just
         * write the page out now */
        if(state->write(ogout.header,1,ogout.header_len,
            out) != (size_t) ogout.header_len) {
                    fprintf(stderr, "Bumming out\n");
          goto cleanup;
                }
        if(state->write(ogout.body,1,ogout.body_len, out) !=
            (size_t) ogout.body_len) {
                    fprintf(stderr, "Bumming out 2\n");
          goto cleanup;
                }
      }
    }
    buffer = ogg_sync_buffer(state->oy, CHUNKSIZE);
    bytes = state->read(buffer,1, CHUNKSIZE, state->in);
    ogg_sync_wrote(state->oy, bytes);
    if(bytes == 0)
    {
      state->eosin = 1;
      break;
    }
  }


cleanup:
  ogg_stream_clear(&streamout);
  ogg_packet_clear(&header_comments);

  free(state->mainbuf);
  free(state->bookbuf);
    state->mainbuf = state->bookbuf = NULL;

  if(!state->eosin)
  {
    state->lasterror =
      _("Error writing stream to output. "
      "Output stream may be corrupted or truncated.");
    return -1;
  }

  return 0;
}



lib/Ogg/Vorbis/Header.pm
package Ogg::Vorbis::Header;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.06';

use Inline C => 'DATA',
  LIBS => '-logg -lvorbis -lvorbisfile',
  INC => '-I/inc',
  AUTO_INCLUDE => '#include "inc/vcedit.h"',
  AUTO_INCLUDE => '#include "inc/vcedit.c"',
  VERSION => '0.06',
  NAME => 'Ogg::Vorbis::Header';

# constructors

# wrap this so $obj->new will work right
sub new {
  my ($id, $path) = @_;
  $id = ref($id) || $id;
  _new($id, $path);
}

sub load {
  my ($id, $path) = @_;
  unless (ref($id)) {
    $id = _new($id, $path);
  }
  return $id unless $id;
  $id->_load_info;
  $id->_load_comments;
  return $id;
}

# A number of the instance methods may be handled with perl code.

sub info {
  my ($self, $key) = @_;
  $self->_load_info unless $self->{INFO};
  if ($key) {
    return $self->{INFO}->{$key};
  }
  return $self->{INFO};
}

sub comment_tags {
  my $self = shift;
  $self->_load_comments unless $self->{COMMENTS};
  return keys %{$self->{COMMENTS}};
}

sub comment {
  my ($self, $key) = @_;
  my $result;
  return undef unless $key;
  $self->_load_comments unless $self->{COMMENTS};
  if (! defined ($result = $self->{COMMENTS}->{$key})) {
    return undef;
  }
  return @{$result};
}

sub add_comments {
  my ($self, @comments) = @_;
  # For now play it safe limit both tag and field to minimal ascii
  # will work on utf8 in field later
  return undef if @comments < 2 or @comments % 2 != 0;
  $self->_load_comments unless $self->{COMMENTS};
  while ($#comments >= 0) {
    my $key = shift @comments;
    $key =~ s/[^\x20-\x3C\x3E-\x7D]//g;
    $key = lc($key);
    my $val = shift @comments;
    $val =~ s/[^\x20-\x7D]//g;
    push @{$self->{COMMENTS}->{$key}}, $val;
  }

  return 1;
}

sub edit_comment {
  my ($self, $key, $value, $num) = @_;
  $num ||= 0;

  return undef unless $key and $value and $num =~ /^\d*$/;
  $self->_load_comments unless $self->{COMMENTS};

  my $comment = $self->{COMMENTS}->{$key};
  return undef unless $comment;
  $value =~ s/[^\x20-\x7D]//g;
  return undef if $num >= @$comment;

  my $result = $comment->[$num];
  $comment->[$num] = $value;

  return $result;
}

sub delete_comment {
  my ($self, $key, $num) = @_;
  $num ||= 0;

  return undef unless $key and $num =~ /^\d*$/;
  $self->_load_comments unless $self->{COMMENTS};

  my $comment = $self->{COMMENTS}->{$key};
  return undef unless $comment;
  return undef if $num >= @$comment;

  my $result = splice @$comment, $num, 1;

  if (@$comment == 0) {
    delete($self->{COMMENTS}->{$key});
  }

  return $result;
}

sub clear_comments {
  my ($self, @keys) = @_;

  $self->_load_comments unless $self->{COMMENTS};
  if (@keys) {
    foreach (@keys) {
      return undef unless $self->{COMMENTS}->{$_};
      delete($self->{COMMENTS}->{$_});
    }
  } else {
    foreach (keys %{$self->{COMMENTS}}) {
      delete($self->{COMMENTS}->{$_});
    }
  }
  return 1;
}

sub path {
  my $self = shift;
  return $self->{PATH};
}

1;
__DATA__

=head1 NAME

Ogg::Vorbis::Header - An object-oriented interface to Ogg Vorbis
information and comment fields.

=head1 SYNOPSIS

  use Ogg::Vorbis::Header;
  my $ogg = Ogg::Vorbis::Header->new("song.ogg");
  while (my ($k, $v) = each %{$ogg->info}) {
    print "$k: $v\n";
  }
  foreach my $com ($ogg->comment_tags) {
    print "$com: $_\n" foreach $ogg->comment($com);
  }
  $ogg->add_comments("good", "no", "ok", "yes");
  $ogg->delete_comment("ok");
  $ogg->write_vorbis;


=head1 DESCRIPTION

This module presents an object-oriented interface to Ogg Vorbis files
which allows user to view Vorbis info and comments and to modify or
add comments.

=head1 CONSTRUCTORS

=head2 C<new ($filename)>

Partially opens an Ogg Vorbis file to ensure it exists and is actually
a Vorbis stream.  It then closes the filehandle.  It does not fill in
the object's data fields.  These fields will be automatically filled
the first time they are accessed using the object's instance methods.
Returns C<undef> if there is a problem opening the file or the file is
not valid Ogg Vorbis.

=head2 C<load ([$filename])>

Opens an Ogg Vorbis file, reads its information, and then closes the
filehandle.  Returns C<undef> if there is a problem opening the file
or the file is not valid Ogg Vorbis.  This is both a constructor and
an instance method.  The filename is required in constructor context,
but should be left out when you call this as an instance method on an
object.  When called as an instance method, it (re)loads the info and
comment data from the file.  This can be used to reset the state of
the object if write_vorbis hasn't been called.  Note that the path
parameter is ignored in instance context.

=head1 INSTANCE METHODS

These methods may be called on actual Header objects, using
the -> operator or indirect objects as you prefer.

=head2 C<info ([$key])>

Returns a reference to a hash containing format information about the
Vorbis file.  Hash fields are: version, channels, rate, bitrate_upper,
bitrate_nominal, bitrate_lower, and bitrate_window, length.  The
bitrate_window value is currently unused by the vorbis codec.  You can
modify the referenced hash if you want, but I wouldn't suggest it.

The optional key parameter allows you to extract a single value from
the internal hash (passed by value, not reference).  If the key is
invalid, C<undef> is returned.

=head2 C<comment_tags ()>

Returns an array holding the key values of each comment field.  You
can then use these values to access specific fields using C<comment>.
This may seem somewhat clunky at first but it will speed up most
programs.  In addition, it makes it easier to support the Ogg Vorbis
comment standard which allows multiple fields with the same key.

=head2 C<comment ($key)>

Returns a list of comments given a key.   If the key does not exist,
returns C<undef>.

=head2 C<add_comments ($key, $value, [$key, $value, ...])>

Adds comments with the given keys and values.  Takes an array of
alternating keys and values as parameters.  Keys and values should be
valid ascii in the range 0x20 - 0x7D and the key should exclude 0x3D
('=').  This is a subset of the Vorbis standard which allows this
range for the key field and all of utf8 for the value field.  This
will be fixed in future a release.

If an odd-length array is passed in the routine will fail and return
C<undef>.  Key and value will be trimmed of characters which do not
match the format requirement.

=head2 C<edit_comment ($key, $value, [$num])>

Edits a given comment field.  The optional num field is used to
differentiate between two comments with the same key.  If no num is
supplied, the first value--as reported by C<comment>--is modified.  If
the key or num are invalid, nothing is done and undef is returned.
If all goes well, the old value is returned.

=head2 C<delete_comment ($key, [$num])>

Deletes the comment given by key.  The optional num value can be used
to specify which comment to delete, given duplicate keys.  Leaving num
out will result in only the first instance being deleted.  Returns
C<undef> if key or num are invalid.  If all goes well, the value of
the deleted comment is returned.

=head2 C<clear_comments ([@keys])>

Deletes all of the comments which match keys in the input array or all
of the comments in the stream if called with no arguments.  Returns
C<undef> if any key is invalid, although all keys in the input array
up until that key will be cleared.  Returns true otherwise.

=head2 C<write_vorbis ()>

Write object to its backing file.  No comment modifications will be
seen in the file until this operation is performed.

=head2 C<path ()>

Returns the path/filename of the file the object represents.

=head1 REQUIRES

Inline::C, libogg, libvorbis, libogg-dev, libvorbis-dev.

=head1 AUTHOR

Dan Pemstein E<lt>dan@lcws.orgE<gt>

=head1 COPYRIGHT

Copyright (c) 2003, Dan Pemstein.  All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at
your option) any later version.  A copy of this license is included
with this module (LICENSE.GPL).

A library for editing Ogg Vorbis comments is distributed with this
library as unmodified source code (inc/vcedit.h, inc/vcedit.c,
inc/i18n.h).  This library is Copyright (c) Michael Smith
<msmith@labyrinth.net.au>.  It is licensed under the GNU Library
General Public License (LGPL).  A copy of this license is included
with this module (inc/LICENSE.LGPL).

=head1 SEE ALSO

L<Ogg::Vorbis::Decoder>, L<Inline::C>, L<Audio::Ao>.

=cut

__C__

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <vorbis/codec.h>
#include <vorbis/vorbisfile.h>

#define BUFFSIZE 512

/* Loads info and length from the stream a fills the object's hash */
void _load_info(SV *obj)
{
  OggVorbis_File vf;
  vorbis_info *vi;
  FILE *fd;
  char *ptr;
  HV *th;
  HV *hash = (HV *) SvRV(obj);

  /* Open the vorbis stream file */
  ptr = (char *) SvIV(*(hv_fetch(hash, "_PATH", 5, 0)));
  if ((fd = fopen(ptr, "rb")) == NULL) {
    perror("Error opening file in Ogg::Vorbis::Header::_load_info\n");
    return;
  }

  if (ov_open(fd, &vf, NULL, 0) < 0) {
    fclose(fd);
    perror("Error opening file in Ogg::Vorbis::Header::_load_info\n");
    return;
  }

  vi = ov_info(&vf, -1);

  th = newHV();
  hv_store(th, "version", 7, newSViv(vi->version), 0);
  hv_store(th, "channels", 8, newSViv(vi->channels), 0);
  hv_store(th, "rate", 4, newSViv(vi->rate), 0);
  hv_store(th, "bitrate_upper", 13, newSViv(vi->bitrate_upper), 0);
  hv_store(th, "bitrate_nominal", 15, newSViv(vi->bitrate_nominal), 0);
  hv_store(th, "bitrate_lower", 13, newSViv(vi->bitrate_lower), 0);
  hv_store(th, "bitrate_window", 14, newSViv(vi->bitrate_window), 0);
  hv_store(th, "length", 6, newSVnv(ov_time_total(&vf, -1)), 0);

  hv_store(hash, "INFO", 4, newRV_noinc((SV *) th), 0);

  ov_clear(&vf);
}

/* Loads the commments from the stream and fills the object's hash */
void _load_comments(SV *obj)
{
  OggVorbis_File vf;
  vorbis_comment *vc;
  FILE *fd;
  HV *th;
  SV *ts;
  AV *ta;
  char *half;
  char *ptr;
  int i;
  HV *hash = (HV *) SvRV(obj);

  /* Open the vorbis stream file */
  ptr = (char *) SvIV(*(hv_fetch(hash, "_PATH", 5, 0)));
  if ((fd = fopen(ptr, "rb")) == NULL) {
    perror("Error opening file in Ogg::Vorbis::Header::_load_comments\n");
    return;
  }

  if (ov_open(fd, &vf, NULL, 0) < 0) {
    fclose(fd);
    perror("Error opening file in Ogg::Vorbis::Header::_load_comments\n");
    return;
  }

  vc = ov_comment(&vf, -1);

  th = newHV();
  for (i = 0; i < vc->comments; ++i) {
    half = strchr(vc->user_comments[i], '=');
    if (half == NULL) {
      warn("Comment \"%s\" missing \'=\', skipping...\n",
            vc->user_comments[i]);
      continue;
    }
    if (! hv_exists(th, vc->user_comments[i],
                half - vc->user_comments[i])) {
      ta = newAV();
      ts = newRV_noinc((SV*) ta);
      hv_store(th, vc->user_comments[i], half - vc->user_comments[i],
        ts, 0);
    } else {
      ta = (AV*) SvRV(*(hv_fetch(th, vc->user_comments[i],
            half - vc->user_comments[i], 0)));
    }
    av_push(ta, newSVpv(half + 1, 0));
  }

  hv_store(hash, "COMMENTS", 8, newRV_noinc((SV *) th), 0);

  ov_clear(&vf);
}

/* Our base object constructor.  Creates a blessed hash. */
SV* _new(char *class, char *path)
{
  /* A few variables */
  FILE *fd;
  char *_path;
  OggVorbis_File vf;

  /* Create our new hash and the reference to it. */
  HV *hash = newHV();
  SV *obj_ref = newRV_noinc((SV*) hash);

  /* Save an internal (c-style) rep of the path */
  _path = strdup(path);
  hv_store(hash, "_PATH", 5, newSViv((IV) _path), 0);

  /* Open the vorbis stream file */
  if ((fd = fopen(path, "rb")) == NULL)
    return &PL_sv_undef;

  if (ov_test(fd, &vf, NULL, 0) < 0) {
    fclose(fd);
    return &PL_sv_undef;
  }

  /* Values stored at base level */
  hv_store(hash, "PATH", 4, newSVpv(path, 0), 0);

  /* Close our OggVorbis_File cause we don't want to keep the file
   * descriptor open.
   */
  ov_clear(&vf);

  /* Bless the hashref to create a class object */
  sv_bless(obj_ref, gv_stashpv(class, FALSE));

  return obj_ref;
}

/* These comment manipulation functions use the vcedit library by
 * Michael Smith.  They also borrow quite a bit from vorbiscomment
 * (vcomment.c) by Michael Smith and Ralph Giles.
 */
int write_vorbis (SV *obj)
{
  vcedit_state *state;
  vorbis_comment *vc;
  char *inpath, *outpath, *key, *val;
  FILE *fd, *fd2, *fd3, *fd4;
  HV *hash = (HV *) SvRV(obj);
  HV *chash;
  AV *vals;
  HE *hval;
  int bytes;
  char buffer[BUFFSIZE];
  I32 i, j, num;


  /* Skip if comments hasn't been opened */
  if (! hv_exists(hash, "COMMENTS", 8)) {
    return 0;
  }

  /* Set up the input and output paths */
  inpath = (char *) SvIV(*(hv_fetch(hash, "_PATH", 5, 0)));
  outpath = malloc(strlen(inpath) + (8 * sizeof(char)));
  strcpy(outpath, inpath);
  strcat(outpath, ".ovitmp");

  /* Open the files */
  if ((fd = fopen(inpath, "rb")) == NULL) {
    perror("Error opening file in Ogg::Vorbis::Header::write\n");
    free(outpath);
    return &PL_sv_undef;
  }

  if ((fd2 = fopen(outpath, "w+b")) == NULL) {
    perror("Error opening temp file in Ogg::Vorbis::Header::write\n");
    fclose(fd);
    free(outpath);
    return &PL_sv_undef;
  }

  /* Setup the state and comments structs */
  state = vcedit_new_state();
  if (vcedit_open(state, fd) < 0) {
    perror("Error opening stream in Ogg::Vorbis::Header::add_comment\n");
    fclose(fd);
    fclose(fd2);
    unlink(outpath);
    free(outpath);
    return &PL_sv_undef;
  }
  vc = vcedit_comments(state);

  /* clear the old comment fields */
  vorbis_comment_clear(vc);
  vorbis_comment_init(vc);

  /* Write the comment fields from the hash
   * FIX: This doesn't preserve order, which may or may not be a problem
   */
  chash = (HV *) SvRV(*(hv_fetch(hash, "COMMENTS", 8, 0)));

  num = hv_iterinit(chash);
  for (i = 0; i < num; ++i) {
    hval = hv_iternext(chash);
    key = SvPV_nolen(hv_iterkeysv(hval));
    vals = (AV*) SvRV(*(hv_fetch(chash, key, strlen(key), 0)));
    for (j = 0; j <= av_len(vals); ++j) {
      val = SvPV_nolen(*av_fetch(vals, j, 0));
      vorbis_comment_add_tag(vc, key, val);
    }
  }

  /* Write out the new stream */
  if (vcedit_write(state, fd2) < 0) {
    perror("Error writing stream in Ogg::Vorbis::Header::add_comment\n");
    fclose(fd);
    fclose(fd2);
    vcedit_clear(state);
    unlink(outpath);
    free(outpath);
    return &PL_sv_undef;
  }

  fclose(fd);
  fclose(fd2);
  vcedit_clear(state);
  if ((fd = fopen(outpath, "rb")) == NULL) {
    perror("Error copying tempfile in Ogg::Vorbis::Header::add_comment\n");
    unlink(outpath);
    free(outpath);
    return &PL_sv_undef;
  }

  if ((fd2 = fopen(inpath, "wb")) == NULL) {
    perror("Error copying tempfile in Ogg::Vorbis::Header::write_vorbis\n");
    fclose(fd);
    unlink(outpath);
    free(outpath);
    return &PL_sv_undef;
  }

  while ((bytes = fread(buffer, 1, BUFFSIZE, fd)) > 0)
    fwrite(buffer, 1, bytes, fd2);

  fclose(fd);
  fclose(fd2);
  unlink(outpath);
  free(outpath);

  return 1;
}

/* We strdup'd the internal path string so we need to free it */
void DESTROY (SV *obj)
{
  char *ptr;
  HV *hash = (HV *) SvRV(obj);

  ptr = (char *) SvIV(*(hv_fetch(hash, "_PATH", 5, 0)));
  free(ptr);
}
